File: java/com/example/spottio/GroupSettingsActivity.java
package com.example.spottio;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.text.InputType;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.bumptech.glide.Glide;
import com.google.firebase.firestore.FieldValue;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.SetOptions;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GroupSettingsActivity extends AppCompatActivity {

    private String currentUser;
    private String groupId;
    private String groupAdminId;
    private boolean isAdmin = false;

    private ImageView ivGroupIcon;
    private TextView tvGroupName, tvGroupDesc, tvEditIconHint;
    private ImageButton btnEditGroupName, btnEditGroupDesc;
    private ListView lvGroupMembers;
    private Button btnAddMember, btnLeaveGroup;

    private FirebaseFirestore db;
    private GroupMemberAdapter memberAdapter;
    private List<String> memberIds;

    private final ActivityResultLauncher<String> pickGroupIcon = registerForActivityResult(
            new ActivityResultContracts.GetContent(), uri -> {
                if (uri != null) {
                    uploadGroupIcon(uri);
                }
            }
    );

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_group_settings);

        SharedPreferences prefs = getSharedPreferences("SpottioPrefs", Context.MODE_PRIVATE);
        currentUser = prefs.getString("username_attivo", null);

        groupId = getIntent().getStringExtra("group_id");
        String passedGroupName = getIntent().getStringExtra("group_name");

        if (groupId == null || currentUser == null) {
            finish();
            return;
        }

        db = FirebaseFirestore.getInstance();
        memberIds = new ArrayList<>();

        ivGroupIcon = findViewById(R.id.ivGroupIcon);
        tvGroupName = findViewById(R.id.tvGroupName);
        tvGroupDesc = findViewById(R.id.tvGroupDesc);
        tvEditIconHint = findViewById(R.id.tvEditIconHint);
        btnEditGroupName = findViewById(R.id.btnEditGroupName);
        btnEditGroupDesc = findViewById(R.id.btnEditGroupDesc);
        lvGroupMembers = findViewById(R.id.lvGroupMembers);
        btnAddMember = findViewById(R.id.btnAddMember);
        btnLeaveGroup = findViewById(R.id.btnLeaveGroup);

        if (passedGroupName != null && !passedGroupName.isEmpty()) {
            tvGroupName.setText(passedGroupName);
        }

        // --- INIZIO FIX: CREAZIONE DEL LISTENER PER L'ADAPTER ---
        GroupMemberAdapter.GroupMemberListener listener = new GroupMemberAdapter.GroupMemberListener() {
            @Override
            public void onRemoveMemberRequest(String memberId) {
                // Il popup ora è generato dall'Activity (Sicuro e senza leak!)
                new AlertDialog.Builder(GroupSettingsActivity.this)
                        .setTitle("Rimuovi partecipante")
                        .setMessage("Vuoi rimuovere " + memberId + " dal gruppo?")
                        .setPositiveButton("Rimuovi", (d, w) -> removeMemberFromGroup(memberId))
                        .setNegativeButton("Annulla", null)
                        .show();
            }

            @Override
            public void onMemberClick(String memberId) {
                // Opzionale: Se vuoi che cliccando un partecipante si apra il suo profilo,
                // puoi implementare l'Intent qui. Altrimenti lascialo vuoto.
            }
        };

        // PASSIAMO IL LISTENER AL COSTRUTTORE (Ora gli argomenti coincidono)
        memberAdapter = new GroupMemberAdapter(this, memberIds, currentUser, db, groupId, listener);
        lvGroupMembers.setAdapter(memberAdapter);
        // --- FINE FIX ---

        loadGroupData();

        btnEditGroupName.setOnClickListener(v -> showEditDialog("Nome", "name", tvGroupName.getText().toString()));
        btnEditGroupDesc.setOnClickListener(v -> showEditDialog("Descrizione", "groupDescription", tvGroupDesc.getText().toString()));
        ivGroupIcon.setOnClickListener(v -> pickGroupIcon.launch("image/*"));

        btnAddMember.setOnClickListener(v -> showAddMemberDialog());
        btnLeaveGroup.setOnClickListener(v -> leaveGroup());
    }

    private void loadGroupData() {
        db.collection("groups").document(groupId).addSnapshotListener((snapshot, error) -> {
            if (error != null || snapshot == null || !snapshot.exists()) {
                checkAdminPermissions();
                return;
            }

            String name = snapshot.getString("name");
            String desc = snapshot.getString("groupDescription");
            String iconUri = snapshot.getString("groupIconUri");
            groupAdminId = snapshot.getString("createdBy");

            List<String> members = (List<String>) snapshot.get("members");

            if (name != null && !name.isEmpty()) tvGroupName.setText(name);

            if (desc != null && !desc.isEmpty()) {
                tvGroupDesc.setText(desc);
            } else {
                tvGroupDesc.setText("Nessuna descrizione");
            }

            if (iconUri != null && !iconUri.isEmpty()) {
                Glide.with(this).load(iconUri).into(ivGroupIcon);
            }

            if (members != null) {
                memberIds.clear();
                memberIds.addAll(members);
                memberAdapter.setAdminId(groupAdminId);
                memberAdapter.notifyDataSetChanged();
            }

            checkAdminPermissions();
        });
    }

    private void checkAdminPermissions() {
        if (groupAdminId == null) {
            isAdmin = true;
        } else {
            isAdmin = currentUser.equals(groupAdminId);
        }

        btnEditGroupName.setVisibility(View.VISIBLE);
        btnEditGroupDesc.setVisibility(View.VISIBLE);
        tvEditIconHint.setVisibility(View.VISIBLE);

        btnAddMember.setVisibility(isAdmin ? View.VISIBLE : View.GONE);
        memberAdapter.setCurrentUserIsAdmin(isAdmin);
    }

    private void showEditDialog(String titolo, String fieldName, String testoAttuale) {
        EditText input = new EditText(this);
        input.setText(testoAttuale.equals("Nessuna descrizione") ? "" : testoAttuale);
        input.setSelection(input.getText().length());

        new AlertDialog.Builder(this)
                .setTitle("Modifica " + titolo)
                .setView(input)
                .setPositiveButton("Salva", (dialog, which) -> {
                    String newText = input.getText().toString().trim();

                    Map<String, Object> data = new HashMap<>();
                    data.put(fieldName, newText);
                    db.collection("groups").document(groupId).set(data, SetOptions.merge());

                    if (fieldName.equals("name")) {
                        Map<String, Object> previewData = new HashMap<>();
                        previewData.put("name", newText);
                        db.collection("chat_previews").document(groupId).set(previewData, SetOptions.merge());
                    }
                })
                .setNegativeButton("Annulla", null)
                .show();
    }

    private void showAddMemberDialog() {
        EditText input = new EditText(this);
        input.setHint("Es: MarioRossi");
        input.setInputType(InputType.TYPE_CLASS_TEXT);

        new AlertDialog.Builder(this)
                .setTitle("Aggiungi Partecipante")
                .setMessage("Inserisci l'username esatto dell'utente:")
                .setView(input)
                .setPositiveButton("Aggiungi", (dialog, which) -> {
                    String username = input.getText().toString().trim();
                    if (!username.isEmpty() && !memberIds.contains(username)) {
                        verificaEAggiungiUtente(username);
                    } else if (memberIds.contains(username)) {
                        Toast.makeText(this, "L'utente è già nel gruppo!", Toast.LENGTH_SHORT).show();
                    }
                })
                .setNegativeButton("Annulla", null)
                .show();
    }

    private void verificaEAggiungiUtente(String username) {
        db.collection("users").document(username).get().addOnSuccessListener(doc -> {
            if (doc.exists()) {
                db.collection("groups").document(groupId).update("members", FieldValue.arrayUnion(username))
                        .addOnSuccessListener(a -> Toast.makeText(this, username + " aggiunto al gruppo!", Toast.LENGTH_SHORT).show());
            } else {
                Toast.makeText(this, "Utente non trovato.", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void leaveGroup() {
        new AlertDialog.Builder(this)
                .setTitle("Abbandona Gruppo")
                .setMessage("Sei sicuro di voler uscire da questo gruppo?")
                .setPositiveButton("Esci", (dialog, which) -> {
                    if (isAdmin && memberIds.size() > 1) {
                        String nuovoAdmin = memberIds.get(0).equals(currentUser) ? memberIds.get(1) : memberIds.get(0);

                        db.collection("groups").document(groupId).update(
                                "members", FieldValue.arrayRemove(currentUser),
                                "createdBy", nuovoAdmin
                        ).addOnSuccessListener(a -> chiudiETornaAllaHome());
                    } else {
                        db.collection("groups").document(groupId).update("members", FieldValue.arrayRemove(currentUser))
                                .addOnSuccessListener(a -> chiudiETornaAllaHome());
                    }
                })
                .setNegativeButton("Annulla", null)
                .show();
    }

    private void chiudiETornaAllaHome() {
        Toast.makeText(this, "Hai abbandonato il gruppo", Toast.LENGTH_SHORT).show();
        Intent intent = new Intent(this, HomeActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(intent);
        finish();
    }

    public void removeMemberFromGroup(String memberId) {
        db.collection("groups").document(groupId)
                .update("members", FieldValue.arrayRemove(memberId))
                .addOnSuccessListener(a -> Toast.makeText(this, "Membro rimosso", Toast.LENGTH_SHORT).show());
    }

    private void uploadGroupIcon(Uri uri) {
        Toast.makeText(this, "Aggiornamento icona...", Toast.LENGTH_SHORT).show();
        String uriString = uri.toString();

        Map<String, Object> iconData = new HashMap<>();
        iconData.put("groupIconUri", uriString);

        db.collection("groups").document(groupId).set(iconData, SetOptions.merge())
                .addOnSuccessListener(a -> {
                    Map<String, Object> previewData = new HashMap<>();
                    previewData.put("profilePicUri", uriString);
                    db.collection("chat_previews").document(groupId).set(previewData, SetOptions.merge());
                    Toast.makeText(this, "Icona aggiornata", Toast.LENGTH_SHORT).show();
                });
    }
}

File: app/src/main/java/com/example/spottio/GroupSettingsActivity.kt
package com.example.spottio

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.InputType
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.ListView
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.bumptech.glide.Glide
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions

class GroupSettingsActivity : AppCompatActivity() {

    private var currentUser: String? = null
    private var groupId: String? = null
    private var groupAdminId: String? = null
    private var isAdmin = false

    private lateinit var ivGroupIcon: ImageView
    private lateinit var tvGroupName: TextView
    private lateinit var tvGroupDesc: TextView
    private lateinit var tvEditIconHint: TextView
    private lateinit var btnEditGroupName: ImageButton
    private lateinit var btnEditGroupDesc: ImageButton
    private lateinit var lvGroupMembers: ListView
    private lateinit var btnAddMember: Button
    private lateinit var btnLeaveGroup: Button

    private lateinit var db: FirebaseFirestore
    private lateinit var memberAdapter: GroupMemberAdapter
    private val memberIds = mutableListOf<String>()

    private val pickGroupIcon = registerForActivityResult(ActivityResultContracts.GetContent()) { uri: Uri? ->
        uri?.let { uploadGroupIcon(it) }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_group_settings)

        val prefs = getSharedPreferences("SpottioPrefs", Context.MODE_PRIVATE)
        currentUser = prefs.getString("username_attivo", null)

        groupId = intent.getStringExtra("group_id")
        val passedGroupName = intent.getStringExtra("group_name")

        if (groupId == null || currentUser == null) {
            finish()
            return
        }

        db = FirebaseFirestore.getInstance()

        ivGroupIcon = findViewById(R.id.ivGroupIcon)
        tvGroupName = findViewById(R.id.tvGroupName)
        tvGroupDesc = findViewById(R.id.tvGroupDesc)
        tvEditIconHint = findViewById(R.id.tvEditIconHint)
        btnEditGroupName = findViewById(R.id.btnEditGroupName)
        btnEditGroupDesc = findViewById(R.id.btnEditGroupDesc)
        lvGroupMembers = findViewById(R.id.lvGroupMembers)
        btnAddMember = findViewById(R.id.btnAddMember)
        btnLeaveGroup = findViewById(R.id.btnLeaveGroup)

        if (!passedGroupName.isNullOrEmpty()) {
            tvGroupName.text = passedGroupName
        }

        // --- CREAZIONE DEL LISTENER PER L'ADAPTER ---
        val listener = object : GroupMemberAdapter.GroupMemberListener {
            override fun onRemoveMemberRequest(memberId: String) {
                AlertDialog.Builder(this@GroupSettingsActivity)
                    .setTitle("Rimuovi partecipante")
                    .setMessage("Vuoi rimuovere $memberId dal gruppo?")
                    .setPositiveButton("Rimuovi") { _, _ -> removeMemberFromGroup(memberId) }
                    .setNegativeButton("Annulla", null)
                    .show()
            }

            override fun onMemberClick(memberId: String) {
                // Opzionale
            }
        }

        memberAdapter = GroupMemberAdapter(this, memberIds, currentUser!!, db, groupId!!, listener)
        lvGroupMembers.adapter = memberAdapter

        loadGroupData()

        btnEditGroupName.setOnClickListener { showEditDialog("Nome", "name", tvGroupName.text.toString()) }
        btnEditGroupDesc.setOnClickListener { showEditDialog("Descrizione", "groupDescription", tvGroupDesc.text.toString()) }
        ivGroupIcon.setOnClickListener { pickGroupIcon.launch("image/*") }

        btnAddMember.setOnClickListener { showAddMemberDialog() }
        btnLeaveGroup.setOnClickListener { leaveGroup() }
    }

    private fun loadGroupData() {
        db.collection("groups").document(groupId!!).addSnapshotListener { snapshot, error ->
            if (error != null || snapshot == null || !snapshot.exists()) {
                checkAdminPermissions()
                return@addSnapshotListener
            }

            val name = snapshot.getString("name")
            val desc = snapshot.getString("groupDescription")
            val iconUri = snapshot.getString("groupIconUri")
            groupAdminId = snapshot.getString("createdBy")

            @Suppress("UNCHECKED_CAST")
            val members = snapshot.get("members") as? List<String>

            if (!name.isNullOrEmpty()) tvGroupName.text = name

            if (!desc.isNullOrEmpty()) {
                tvGroupDesc.text = desc
            } else {
                tvGroupDesc.text = "Nessuna descrizione"
            }

            if (!iconUri.isNullOrEmpty()) {
                Glide.with(this).load(iconUri).into(ivGroupIcon)
            }

            if (members != null) {
                memberIds.clear()
                memberIds.addAll(members)
                memberAdapter.setAdminId(groupAdminId)
                memberAdapter.notifyDataSetChanged()
            }

            checkAdminPermissions()
        }
    }

    private fun checkAdminPermissions() {
        isAdmin = if (groupAdminId == null) {
            true
        } else {
            currentUser == groupAdminId
        }

        btnEditGroupName.visibility = View.VISIBLE
        btnEditGroupDesc.visibility = View.VISIBLE
        tvEditIconHint.visibility = View.VISIBLE

        btnAddMember.visibility = if (isAdmin) View.VISIBLE else View.GONE
        memberAdapter.setCurrentUserIsAdmin(isAdmin)
    }

    private fun showEditDialog(titolo: String, fieldName: String, testoAttuale: String) {
        val input = EditText(this)
        input.setText(if (testoAttuale == "Nessuna descrizione") "" else testoAttuale)
        input.setSelection(input.text.length)

        AlertDialog.Builder(this)
            .setTitle("Modifica $titolo")
            .setView(input)
            .setPositiveButton("Salva") { _, _ ->
                val newText = input.text.toString().trim()

                val data = hashMapOf<String, Any>(fieldName to newText)
                db.collection("groups").document(groupId!!).set(data, SetOptions.merge())

                if (fieldName == "name") {
                    val previewData = hashMapOf<String, Any>("name" to newText)
                    db.collection("chat_previews").document(groupId!!).set(previewData, SetOptions.merge())
                }
            }
            .setNegativeButton("Annulla", null)
            .show()
    }

    private fun showAddMemberDialog() {
        val input = EditText(this)
        input.hint = "Es: MarioRossi"
        input.inputType = InputType.TYPE_CLASS_TEXT

        AlertDialog.Builder(this)
            .setTitle("Aggiungi Partecipante")
            .setMessage("Inserisci l'username esatto dell'utente:")
            .setView(input)
            .setPositiveButton("Aggiungi") { _, _ ->
                val username = input.text.toString().trim()
                if (username.isNotEmpty() && !memberIds.contains(username)) {
                    verificaEAggiungiUtente(username)
                } else if (memberIds.contains(username)) {
                    Toast.makeText(this, "L'utente è già nel gruppo!", Toast.LENGTH_SHORT).show()
                }
            }
            .setNegativeButton("Annulla", null)
            .show()
    }

    private fun verificaEAggiungiUtente(username: String) {
        db.collection("users").document(username).get().addOnSuccessListener { doc ->
            if (doc.exists()) {
                db.collection("groups").document(groupId!!).update("members", FieldValue.arrayUnion(username))
                    .addOnSuccessListener { Toast.makeText(this, "$username aggiunto al gruppo!", Toast.LENGTH_SHORT).show() }
            } else {
                Toast.makeText(this, "Utente non trovato.", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun leaveGroup() {
        AlertDialog.Builder(this)
            .setTitle("Abbandona Gruppo")
            .setMessage("Sei sicuro di voler uscire da questo gruppo?")
            .setPositiveButton("Esci") { _, _ ->
                if (isAdmin && memberIds.size > 1) {
                    val nuovoAdmin = if (memberIds[0] == currentUser) memberIds[1] else memberIds[0]

                    db.collection("groups").document(groupId!!).update(
                        "members", FieldValue.arrayRemove(currentUser),
                        "createdBy", nuovoAdmin
                    ).addOnSuccessListener { chiudiETornaAllaHome() }
                } else {
                    db.collection("groups").document(groupId!!).update("members", FieldValue.arrayRemove(currentUser))
                        .addOnSuccessListener { chiudiETornaAllaHome() }
                }
            }
            .setNegativeButton("Annulla", null)
            .show()
    }

    private fun chiudiETornaAllaHome() {
        Toast.makeText(this, "Hai abbandonato il gruppo", Toast.LENGTH_SHORT).show()
        val intent = Intent(this, HomeActivity::class.java).apply {
            addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
        }
        startActivity(intent)
        finish()
    }

    fun removeMemberFromGroup(memberId: String) {
        db.collection("groups").document(groupId!!)
            .update("members", FieldValue.arrayRemove(memberId))
            .addOnSuccessListener { Toast.makeText(this, "Membro rimosso", Toast.LENGTH_SHORT).show() }
    }

    private fun uploadGroupIcon(uri: Uri) {
        Toast.makeText(this, "Aggiornamento icona...", Toast.LENGTH_SHORT).show()
        val uriString = uri.toString()

        val iconData = hashMapOf<String, Any>("groupIconUri" to uriString)

        db.collection("groups").document(groupId!!).set(iconData, SetOptions.merge())
            .addOnSuccessListener {
                val previewData = hashMapOf<String, Any>("profilePicUri" to uriString)
                db.collection("chat_previews").document(groupId!!).set(previewData, SetOptions.merge())
                Toast.makeText(this, "Icona aggiornata", Toast.LENGTH_SHORT).show()
            }
    }
}